etcd_3_6: init 3.6.4

superherointj opened the pull request:
etcd_3_6: init 3.6.4

Release: https://github.com/etcd-io/etcd/releases/tag/v3.6.4

Changelog: https://github.com/etcd-io/etcd/blob/main/CHANGELOG/CHANGELOG-3.6.md

Migration notes: https://etcd.io/docs/v3.6/upgrades/upgrade_3_6/

Known future/pending work:
- Etcd 3.6 tests needs work and is pending. (I don't have a solution for it yet.)
  - Updating latest etcd version should happen after etcd_3_6 is fully tested.

CC @dtomvan

acid-bong commented on pkgs/top-level/all-packages.nix:
```diff
@@ -10138,6 +10138,9 @@ with pkgs;
   etcd_3_4 = callPackage ../servers/etcd/3.4.nix { };
   etcd_3_5 = callPackage ../servers/etcd/3.5 { };
+  etcd_3_6 = callPackage ../servers/etcd/3_6 {
+    buildGoModule = buildGo123Module;
```
Why is a newer app version compiled with an older Go version (currently Go-1.24 is pinned)?

superherointj replied:
https://github.com/etcd-io/etcd/blob/v3.6.4/.go-version

superherointj replied:
> Why is a newer app version compiled with an older Go version (currently Go-1.24 is pinned)?

Older versions have the pinning outside the top-level.
Because of override interface it's being corrected in two different PRs:
- https://github.com/NixOS/nixpkgs/pull/430176
- https://github.com/NixOS/nixpkgs/pull/430420

superherointj replied:
Upstream requires that specific Go version:
https://github.com/etcd-io/etcd/blob/v3.6.4/.go-version

Older versions have the pinning outside the top-level.
Because of override interface it's being corrected in two different PRs:
- https://github.com/NixOS/nixpkgs/pull/430176
- https://github.com/NixOS/nixpkgs/pull/430420

Thanks for reviewing the PR.

superherointj replied:
Upstream requires that specific Go version:
https://github.com/etcd-io/etcd/blob/v3.6.4/.go-version

Older versions have the pinning outside the top-level, such as:
https://github.com/NixOS/nixpkgs/blob/master/pkgs/servers/etcd/3.5/default.nix#L3

Override interface is being fixed in these 2 PRs:
- https://github.com/NixOS/nixpkgs/pull/430176
- https://github.com/NixOS/nixpkgs/pull/430420

Thanks for reviewing the PR.

dtomvan reviewed (commented):
Nits for days (I understand most of those are also applicable to other etcd derivations so I guess they are best addressed once we do introduce a `common.nix` or the like). LGTM ultimately, I'll look at it again once tests are in place I guess. Yes, you can add me as maintainer, but I was under the impression that @offlinehacker is also an etcd maintainer?

superherointj commented on pkgs/servers/etcd/3_6/default.nix:
```diff
@@ -0,0 +1,114 @@
+
+  meta = {
+    description = "Distributed reliable key-value store for the most critical data of a distributed system";
+    downloadPage = "https://github.com/etcd-io/etcd";
```
The `downloadPage` is the place the software is downloaded from. (GitHub in this case) I mainly use this copy and paste to visit the downloadPage.
The homepage of `etcd` would be: `https://etcd.io/`

superherointj commented on pkgs/servers/etcd/3_6/default.nix:
```diff
@@ -0,0 +1,114 @@
+    deps = {
+      inherit etcdserver etcdutl etcdctl;
+    };
+    # Fix-Me: Tests for etcd 3.6 needs work.
```
In practice, it just needs work. Last time I tried to make tests work I had issues with the way things are currently structured. I currently don't have the time to properly structure tests. My plan was to sort tests in another future PR.

superherointj commented on pkgs/servers/etcd/3_6/update.sh:
```diff
@@ -0,0 +1,74 @@
+    getAndSetVendorHash etcdUtlVendorHash etcdutl
+    getAndSetVendorHash etcdCtlVendorHash etcdctl
+
+    # `git` flag here is to be used by local maintainers to speed up the bump process
+    if [ $# -eq 1 ] && [ "$1" = "git" ]; then
+        git switch -c "package-$ETCD_PKG_NAME-$LATEST_VERSION"
+        git add "$ETCD_PATH"/default.nix
+        git commit -m "$ETCD_PKG_NAME: $OLD_VERSION -> $LATEST_VERSION
+
+Release: https://github.com/etcd-io/etcd/releases/tag/$LATEST_TAG"
+    fi
+
```
Are you sure this uses "mic92/nix-update"? My understanding is this is a custom script.

AFAIK the reason this is a custom script is because etcd has multiple components with a lifecycle of it's own.

I believe what you are referencing is a helper for maintainers for locally creating git commit messages automatically. This is not the same use case from "mic92/nix-update" because this is used apart from that. I usually use this by direct execution of script. The other etcd 3.5 update script is very like this.

superherointj commented:
> Nits for days (I understand most of those are also applicable to other etcd derivations so I guess they are best addressed once we do introduce a `common.nix` or the like).

`etcd` upkeep in nixpkgs is delayed/lacking. If you check history, you'll see I have done lots of upkeep for it. But I don't aim for perfection because I don't have the time/priority for sorting everything. I try to move things forward as much as possible, particularly the most critical parts. By now I've given up on the expectation of perfection. As usual, the wants are many, but the capabilities are limited. And I agree, update script, tests, and even package could be abstracted. Also, other things.

If I did not upstream this PR without tests, what would help is eventual code drift. Then, next time I'd have to sort the code drift instead of tackling tests, or something else. I'd rather do multiple iterations of contributions than try to sort everything at once and fail. The review also happens to be easier. The PR is incremental since most is a copy and paste of what already exist. The priority would be to have the package available and working. Whenever clean-up happens, it would happen from having a solid working reference which is better.

dtomvan commented:
You are right. Of course. I just like to point out things I see when I look at something carefully. When I say "nit" I dont't expect anyone to do something about it. You might call it pointless to even leave the comment, but I guess there's a little bit of OCD if I don't point things out.... 😅 I hope you understand.

When it comes to initting a leaf package most of the time, if it works it works... but then we should indeed recognize that improvements can always be made.

That said, this PR looks good as long as we can still bypass `nixpkgs-vet`. Committers can't just merge PR's normally anymore when the "no PR failures" check is red...

superherointj commented:
@SuperSandro2000 Do you know how to pass `nixpkgs-vet` here?

acid-bong commented:
> how to pass `nixpkgs-vet`

only by moving the package to `pkgs/by-name`, but it's not necessary at the moment (it's better to move all etcd instances simultaneously, but that's out of this PR's scope). your PR is mergeable anyway

dtomvan commented on pkgs/servers/etcd/3_6/default.nix:
```diff
@@ -0,0 +1,134 @@
+    downloadPage = "https://github.com/etcd-io/etcd";
+    license = lib.licenses.asl20;
+    homepage = "https://etcd.io/";
+    maintainers = with lib.maintainers; [ dtomvan ];
```
I see... You have 10x the PRs merged compared to me anyways. So I am sorry that that interaction happened to you. I don't see maintainer status as a big honor or something, just that you get pinged because you opted in (and also that you claim to, well, _maintain_ the package 😅 ). Well... I guess you are the first unwritten maintainer of nixpkgs packages...

Funnily enough, the person who got angry at you for adding yourself as maintainer to "his package" hasn't opened a PR since...

superherointj committed:
etcd_3_6: init 3.6.4

Release: https://github.com/etcd-io/etcd/releases/tag/v3.6.4

Changelog: https://github.com/etcd-io/etcd/blob/main/CHANGELOG/CHANGELOG-3.6.md

Migration notes: https://etcd.io/docs/v3.6/upgrades/upgrade_3_6/

superherointj commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 430436`
Commit: `6436f6d30c209b3594ddc518efc1d15bda1a870f`

---
### `aarch64-darwin`
<details>
  <summary>:x: 1 package failed to build:</summary>
  <ul>
    <li>etcd</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>etcd_3_5</li>
  </ul>
</details>

superherointj commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 430436`
Commit: `487b0a929c5f55caa850a1e5557f3bd60cf1a503`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 2 packages built:</summary>
  <ul>
    <li>etcd</li>
    <li>etcd_3_5</li>
  </ul>
</details>

katexochen reviewed (approved):
Thanks!
